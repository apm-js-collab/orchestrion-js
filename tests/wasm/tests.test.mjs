import { create } from "../../pkg/orchestrion_js.js";
import { describe, test, expect } from "vitest";

describe('Orchestrion JS Transformer', () => {
    const instrumentor = create([
        {
            channelName: "up:constructor",
            module: { name: "one", versionRange: ">=1", filePath: "index.js" },
            functionQuery: { className: "Up" },
        },
        {
            channelName: "up:fetch",
            module: { name: "one", versionRange: ">=1", filePath: "index.js" },
            functionQuery: {
                className: "Up",
                methodName: "fetch",
                kind: "Sync",
            },
        },
    ]);

    const matchedTransforms = instrumentor.getTransformer(
        "one",
        "1.0.0",
        "index.js",
    );

    test('should get transformer for matching module', () => {
        expect(matchedTransforms).toBeTruthy();
    });

    test('should transform ESM module correctly', () => {
        const originalEsm = `export class Up {
	constructor() {
		console.log('constructor')
	}

	fetch() {
		console.log('fetch')
	}
}`;

        const output = matchedTransforms.transform(originalEsm, 'esm');
        expect(output).toMatchSnapshot();
    });

    test('should transform CommonJS module correctly', () => {
        const originalCjs = `module.exports = class Up {
	constructor() {
		console.log('constructor')
	}

	fetch() {
		console.log('fetch')
	}
}

`;
        const outputCjs = matchedTransforms.transform(originalCjs, 'cjs');
        expect(outputCjs).toMatchSnapshot();
    });

    test('should transform TypeScript with source map correctly', () => {
        const originalTypescript = `export class Up {
    constructor() {
        console.log('constructor');
    }
    fetch(url) {
        console.log('fetch');
    }
}`;

        const originalTypescriptSourceMap = `{"version":3,"file":"typescript.js","sourceRoot":"","sources":["typescript.ts"],"names":[],"mappings":"AAAA,MAAM,OAAO,EAAE;IACd;QACC,OAAO,CAAC,GAAG,CAAC,aAAa,CAAC,CAAA;IAC3B,CAAC;IAED,KAAK,CAAC,GAAW;QAChB,OAAO,CAAC,GAAG,CAAC,OAAO,CAAC,CAAA;IACrB,CAAC;CACD"}`;

        const outputTs = matchedTransforms.transformWithSourcemap(
            "typescript.js",
            originalTypescript,
            "esm",
            originalTypescriptSourceMap,
        );

        expect(outputTs.code).toMatchSnapshot();
        expect(outputTs.map).toMatchSnapshot();
    });

    test('should throw error when no injection points are found', () => {
        const noMatchSource = `export class Down {
	constructor() {
		console.log('constructor')
	}

	fetch() {
		console.log('fetch')
	}
}`;

        expect(() => {
            matchedTransforms.transform(noMatchSource, 'unknown');
        }).toThrow('Failed to find injection points for: ["constructor", "fetch"]');
    });
});
