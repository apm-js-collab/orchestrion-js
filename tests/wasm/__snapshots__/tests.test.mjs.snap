// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Orchestrion JS Transformer > should transform CommonJS module correctly 1`] = `
"const { tracingChannel: tr_ch_apm_tracingChannel } = require("diagnostics_channel");
const tr_ch_apm$up_fetch = tr_ch_apm_tracingChannel("orchestrion:one:up:fetch");
const tr_ch_apm$up_constructor = tr_ch_apm_tracingChannel("orchestrion:one:up:constructor");
module.exports = class Up {
    constructor(){
        const tr_ch_apm_ctx$up_constructor = {
            arguments
        };
        try {
            if (tr_ch_apm$up_constructor.hasSubscribers) {
                tr_ch_apm$up_constructor.start.publish(tr_ch_apm_ctx$up_constructor);
            }
            console.log('constructor');
        } catch (tr_ch_err) {
            if (tr_ch_apm$up_constructor.hasSubscribers) {
                tr_ch_apm_ctx$up_constructor.error = tr_ch_err;
                try {
                    tr_ch_apm_ctx$up_constructor.self = this;
                } catch (refErr) {}
                tr_ch_apm$up_constructor.error.publish(tr_ch_apm_ctx$up_constructor);
            }
            throw tr_ch_err;
        } finally{
            if (tr_ch_apm$up_constructor.hasSubscribers) {
                tr_ch_apm_ctx$up_constructor.self = this;
                tr_ch_apm$up_constructor.end.publish(tr_ch_apm_ctx$up_constructor);
            }
        }
    }
    fetch() {
        const traced = ()=>{
            console.log('fetch');
        };
        if (!tr_ch_apm$up_fetch.hasSubscribers) return traced();
        return tr_ch_apm$up_fetch.traceSync(traced, {
            arguments,
            self: this
        });
    }
};
"
`;

exports[`Orchestrion JS Transformer > should transform ESM module correctly 1`] = `
"import { tracingChannel as tr_ch_apm_tracingChannel } from "diagnostics_channel";
const tr_ch_apm$up_fetch = tr_ch_apm_tracingChannel("orchestrion:one:up:fetch");
const tr_ch_apm$up_constructor = tr_ch_apm_tracingChannel("orchestrion:one:up:constructor");
export class Up {
    constructor(){
        const tr_ch_apm_ctx$up_constructor = {
            arguments
        };
        try {
            if (tr_ch_apm$up_constructor.hasSubscribers) {
                tr_ch_apm$up_constructor.start.publish(tr_ch_apm_ctx$up_constructor);
            }
            console.log('constructor');
        } catch (tr_ch_err) {
            if (tr_ch_apm$up_constructor.hasSubscribers) {
                tr_ch_apm_ctx$up_constructor.error = tr_ch_err;
                try {
                    tr_ch_apm_ctx$up_constructor.self = this;
                } catch (refErr) {}
                tr_ch_apm$up_constructor.error.publish(tr_ch_apm_ctx$up_constructor);
            }
            throw tr_ch_err;
        } finally{
            if (tr_ch_apm$up_constructor.hasSubscribers) {
                tr_ch_apm_ctx$up_constructor.self = this;
                tr_ch_apm$up_constructor.end.publish(tr_ch_apm_ctx$up_constructor);
            }
        }
    }
    fetch() {
        const traced = ()=>{
            console.log('fetch');
        };
        if (!tr_ch_apm$up_fetch.hasSubscribers) return traced();
        return tr_ch_apm$up_fetch.traceSync(traced, {
            arguments,
            self: this
        });
    }
}
"
`;

exports[`Orchestrion JS Transformer > should transform TypeScript with source map correctly 1`] = `
"import { tracingChannel as tr_ch_apm_tracingChannel } from "diagnostics_channel";
const tr_ch_apm$up_fetch = tr_ch_apm_tracingChannel("orchestrion:one:up:fetch");
const tr_ch_apm$up_constructor = tr_ch_apm_tracingChannel("orchestrion:one:up:constructor");
export class Up {
    constructor(){
        const tr_ch_apm_ctx$up_constructor = {
            arguments
        };
        try {
            if (tr_ch_apm$up_constructor.hasSubscribers) {
                tr_ch_apm$up_constructor.start.publish(tr_ch_apm_ctx$up_constructor);
            }
            console.log('constructor');
        } catch (tr_ch_err) {
            if (tr_ch_apm$up_constructor.hasSubscribers) {
                tr_ch_apm_ctx$up_constructor.error = tr_ch_err;
                try {
                    tr_ch_apm_ctx$up_constructor.self = this;
                } catch (refErr) {}
                tr_ch_apm$up_constructor.error.publish(tr_ch_apm_ctx$up_constructor);
            }
            throw tr_ch_err;
        } finally{
            if (tr_ch_apm$up_constructor.hasSubscribers) {
                tr_ch_apm_ctx$up_constructor.self = this;
                tr_ch_apm$up_constructor.end.publish(tr_ch_apm_ctx$up_constructor);
            }
        }
    }
    fetch(url) {
        const traced = ()=>{
            console.log('fetch');
        };
        if (!tr_ch_apm$up_fetch.hasSubscribers) return traced();
        return tr_ch_apm$up_fetch.traceSync(traced, {
            arguments,
            self: this
        });
    }
}
"
`;

exports[`Orchestrion JS Transformer > should transform TypeScript with source map correctly 2`] = `"{"version":3,"file":"typescript.js","sources":["typescript.ts"],"sourceRoot":"","names":[],"mappings":";;;AAAA,MAAM,CAAA,MAAO,EAAE;IACd,aAAA;;;;;;;;YACC,OAAO,CAAC,GAAG,CAAC,aAAa,CAAC,CAAA;;;;;;;;;;;;;;;;IAC3B,CAAC;IAED,KAAK,CAAC,GAAW,EAAA;;YAChB,OAAO,CAAC,GAAG,CAAC,OAAO,CAAC,CAAA;;;;;;;IACrB,CAAC;CACD"}"`;
